Introduce CDI for third-party storage (VM Image/root disk)

Vicente-Cheng opened the pull request:
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
To support third-party storage for VM Image/VM root disk

**Solution:**
Add new VMImage backend and VM root disk mechanism for third-party storage.

**Related Issue:**
https://github.com/harvester/harvester/issues/1199

**Test plan:**
- Try to create VM Image with 3rd-party storage (e.g. LVM)
  - Verify download/upload path
- Try to create VM with the above VM image
- ensure the VM Image PVC would have the annotation `harvesterhci.io/goldenImage: "true"`
- ensure this annotation cannot be modified/deleted
- ensure this volume cannot be removed directly (means you need to remove VMimage to remove the golden image PVC)
- ensure the golden image cannot be used with VM disks

For the download path, we should do the following steps until UI change.
1. apply the following yaml file
```
apiVersion: harvesterhci.io/v1beta1
kind: VirtualMachineImageDownloader
metadata:
  name: <image name> NOTE: not display name
  namespace: <image namespace>
spec:
  imageName: <image name>
```

2. Wait until the above CR is ready, and click the download.

Vicente-Cheng committed:
apis/VMImage: add new field to Virtualmachineimages CRD

    - Add a new field, TargetStorageClassName, which means the storage
      provisioner that will be used to create the VM Image.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
vendor: update the CDI related packages

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
common: make generate

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
VMImage/CDI: Add CDI backend

    - rewrite the `GetImageStorageClassName` func to get the correct
      StorageClassName.
    - Support VM creation with Upload/Download

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
controller/vm: add root disk mechnaism for the 3rd-party storage

    - we will create root disk based on the golden image through
      CDI capability

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
webhook: add golden image related webhook

    - protect the golden image annotation
    - prevent golden image volume from being used for VM

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
controller: add new storage profile controller

    - CDI could use an individual storage profile with a
      specific StorageClass. That will help define the
      different Strategy/SnapshotClass that we want.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
VMImage/CDI: support VM Image download

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

ihcsim reviewed (commented):
Here's the first set of comments and questions. More to come later. There's a lot in here. Maybe in the future, we can break it down into multiple smaller PRs for easier reviews and testing.

Vicente-Cheng committed:
VMImag/CDI: support export volume to VM Image

    - Add CDI common libs

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
controller/pvc: add PVC finalizer

    - Because we use DataVolume to create root disk (PVC), we need to
      remove the DataVolume before PVC deleting. Or the PVC will be
      recreated by CDI controller.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
api/vm: add new checking for backup action

    - Now the CDI volume did not support backup. If VM use any disk
      created by DataVolume, we should disable the backup action.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
webhook: add datavolume validator

    - protect the datavolume owned by VM or VM Image

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
upgrade: add rke2 config file for upgrade

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
test/integration: extend the maintenance timeout to 300 seconds

    - 120 seconds sometime is not enough for node drain

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

WebberHuang1118 reviewed (commented):
LGTM, generally, most of the comments are similar to those at https://github.com/harvester/harvester/pull/7640#pullrequestreview-2640327855. Considering our schedule, we may defer this to the discussion at https://github.com/harvester/harvester/issues/7524.

In the meantime I'm building my env, there are a few questions:
- How the CR `StorageProfile` is created? and what's the default `CloneStrategy`, `clone` or `snapshot`?
- During a new PVC is created from the golden image with `snapshot` strategy, how the related `snapshotClass` is specified for each storage provider?

Thanks

Vicente-Cheng committed:
image/cdi: use vm image operator to get/update

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

ihcsim reviewed (commented):
I got as far as the uploader today. Some comments and questions below. Thanks.

Vicente-Cheng committed:
api/vm: improve the create vm template

    - to support create vm template with 3rd-partry storage solution

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
api: improve the behavior of VM Image Download

    - Now we have the new CRD virtualmachineimagedownloader to create
      the server for vm image download

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
apis: add new CRD VirtualMachineImageDownloader

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng committed:
api/image: add downloadcancel for CDI image

    - Add `downloadcancel` to remove the download server of CDI VM Image

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

ihcsim reviewed (approved):
@Vicente-Cheng Overall, LGTM - thanks! I think we should get this PR merged soon to get QA started. We can handle any subsequent changes or bug fixes in smaller follow-up PRs, instead of churning through this big one.

Vicente-Cheng committed:
charts: add cdi

    - The chart partial comes from oci://registry.suse.com/edge/cdi-chart
    - Update labels of cdi-operator for both Role/RoleBinding, use
      cdi-operator.label macro
    - Update to the latest cdi version v1.61.0

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

ibrokethecloud reviewed (approved):
lgtm. thanks for the enhacement.

bk201 reviewed (approved):
LGTM. Verified with NFS and Rook csi.

WebberHuang1118 reviewed (approved):
LGTM, checked with hostpath-csi.

Please help to answer it https://github.com/harvester/harvester/pull/7640#pullrequestreview-2640371916 when you have time, thanks.
